### PR TITLE
Adding support for non-empty fields (FieldNotEmptyAttribute) and T4 Test Generator Cleanup

### PR DESCRIPTION
**Pull request by jbparker**
Hi @MarcosMeli @mcavigelli
I started writing support on my side but then figured it would be just as well to do it in the library (non-empty fields).  I'm open to building on this more if you had any ideas.

Also, while adding the tests for this I found the T4 generated style of tests (awesome!) but found that the auto-generated code had to be "massaged" quite a bit to be in its current state, so I added proper spacing support to FluentCode.cs.

In that same file, I also removed the duplicate FluentCode.cs and added an assembly reference to the T4 template so it could be used within the T4 template.

Eventually, I could see creating another project for FluentCode.cs (and other code generation throughout the project) so as to not have a potential dependency loop.  However, as long as the project is built now, the T4 transform will work properly.

Again, open to any changes of course and thanks for the product!

### CODE REVIEW

**Comment by jbparker** [comment]
Just wanted to note that I can split up the T4 template cleanup into its own pull request if that helps - was thinking it would be nice to add the  FileHelpers.Tests/Data/GenerateTestData.tt transform on pre-build, but to do that there would need to be a project for code generation (for compiling FluentCode.cs).  Let me know if that helps with this one.

File: FileHelpers/Fields/FieldBase.cs [hunk 1]
```diff
@@ -374,19 +377,6 @@ internal static string AutoPropertyName(FieldInfo fi)
         /// <param name="fi">Field Info Object</param>
         internal FieldBase(FieldInfo fi)
         {
-            IsNullableType = false;
```
Comment (MarcosMeli) [inline FileHelpers/Fields/FieldBase.cs]: I used the trivial init of fields to remember to initialize all fields  (anyway are removed by compiler optimizer later)

Would be better to left all the fields initialized

**Comment by MarcosMeli** [comment]
Wow Great job Jordan

I just review the code and all looks perfect for a Merge

I will add some notes to the diff as a remember to revert some trivial changes.

Lets review that and Merge it

THANKS A TON !!!

File: FileHelpers/Fields/FieldBase.cs (hunk 1 above)
Comment (jbparker) [inline FileHelpers/Fields/FieldBase.cs]: You got it - I'll add that back.  Thanks @MarcosMeli!

**Comment by MarcosMeli** [comment]
About the T4 Generation, I was working in the FileHelpers.Samples project that use T4 to generate HTML Docs for lots of examples.

Also generates an app that navigates for the examples, and compile check the examples

image

**Comment by MarcosMeli** [comment]
Thinking back about the new FieldIsNotEmpty, that particular attribute is for validation purpose, what do you think to create a naming rule for these kind of attributes  ?

For example

FieldValidateIsNotEmpty
ValidateFieldIsNotEmpty

Doing so, we can create a ValidationBaseAttribute and allow user to extend the lib with new attributes

Cheers

**Comment by jbparker** [comment]
Hey @MarcosMeli 
I was thinking the same thing and wanted to get my feet wet with the library - I will definitely start working on that - I'm working on an app for importing files for DB insert so this would be a perfect thing to have (and make my life a lot easier) :+1: 

Sounds good on the T4 - I saw that one and the one for messages - I will probably just move FluentCode.cs to the template then so we don't have to create a project just for that.  Thanks for the help with this!

**Pull request merged**